Add more tests for clickhouse-disks with plain_rewritable

thraeka opened the pull request:
<!---AI changelog entry and formatting assistance: false-->
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Add more tests for clickhouse-disks with plain_rewritable

thraeka committed:
Updated disk app to
Modified existsDirectory to handle root and partial paths
	- Created new existsPartialOrFullPath
		- existsLocalPath is used for file operations and file operations require full path
		- getRemotePathInfoIfExists returns RemotePathInfo and those only exist for full path
	- getRelativeFromRoot and InMemoryDirectoryPathMap issue
		- InMemoryDirectoryPathMap stores entire path without a ‘/’ at the end of the path
		- getRelativeFromRoot returns a directory path with a ‘/’ at the end of the path
		- Results in full paths from getRelativeFromRoot not being found in InMemoryDirectoryMap
		- To resolve this conflict, MetadataStorageFromPlainRewritableObjectStorage::existsDirectory removes ‘/’ at end if path has
Modified iterateSubdirectories
	- Created set so that duplicates are not added to callback
	- Looks for ‘/’ or subdirectory.end for subdirectories (last subdirectory does not end in a ‘/’)

thraeka committed:
Reverted existsDirectory for PROS + overloaded metadata existsDirectory
Goal is to emulate directory for plain rewritable object storage (PROS)
Constraints:
- PROS require full path for file operations (mv, cp, rm)
- to emulate subdirectories for path nav (ls), need intermed paths
- DiskClient utilizes IDisk::existsDirectory in both (mv, cp, rm) and ls
	- modifying existsDirectory for PROB disk to allow intermed paths will affect mv, cp, and rm
	- isDirectory is ideal place to modify

Potential Soln in isDirectory() modification:
- To avoid modifying IDisk::existsDirectory for non-PROS disk, need specific case for PROS in isDirectory()
- Soln 1 (follow disk flow): overload IDisk::existsDirectory(string path, bool emulate_directories)
	- excessive for very specific case of PROS disk
- Soln 2 (break flow and use metadata): overload iMetadataStorage::existsDirectory(string path, bool emulate_directories)
	- requires DiskClient to be aware of metadata fn (add metadata header)
- Soln 3 (break flow and call PROS fn): cast metadata to PROS and use PROS::existsDirectory(string path, bool emulate_directories)
	- This way no need to modify iMetadataStorage base class, but this requires DiskClient to be more PROS aware

Opted for Soln 2

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=86780) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=86780) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=86780) it.</sub>

azat commented:
@thraeka thank you! Can you also please add a test? You can take a look at the `tests/queries/0_stateless/02802_clickhouse_disks_s3_copy.sh` as a reference

thraeka committed:
Added stateless tests for clickhouse-disks ls cmd

thraeka committed:
Updated changelog category to Improvement only

thraeka committed:
check for disk metadata in isDirectory and update test
- some disks don't have metadata so check for metadata
- in test, separate folder creation query

thraeka committed:
fixed flakey test with unique test names

thraeka committed:
updated existsDirectory(const std::string)
- now uses path_map->existsPartialOrFullPath
- deleted existsDirectory(const std::string, bool) and reverted files

thraeka committed:
Prev CI passed all test, but error ++. Rerunning CI

thraeka committed:
Another try for CI run

azat reviewed (commented):
Thank you, it looks good to me, please take a look at the comments

thraeka committed:
Updated per request

thraeka commented on src/Disks/ObjectStorages/InMemoryDirectoryPathMap.h:
```diff
@@ -210,22 +229,20 @@ class InMemoryDirectoryPathMap
+            std::string subdirectory = std::string(start_of_subdirectory, end_of_subdirectory);
+            subdirectory.push_back('/');
+            if (subdirectories.insert(subdirectory).second)
+                callback(subdirectory);
```
Yes. There are ls commands in the test script

azat replied:
Hm, I still don't see a test for this case, and I run the test w/o your changes and it passes

thraeka committed:
added test case for returning directories only once

thraeka commented:
> @thraeka can you also PTAL why the following does not work (this is with your PR applied up to [c8051b1](https://github.com/ClickHouse/ClickHouse/commit/c8051b1f57e4292268e2a323bdf4b084dc497f92))
> 
> ```
> $ clickhouse-disks -C /src/ch/clickhouse/programs/server/config.xml --disk s3_plain_rewritable_disk
> s3_plain_rewritable_disk:/$ ls
> s3_plain_rewritable_disk:/$ ls /
> s3_plain_rewritable_disk:/$ touch /foo/bar/bam/baz
> s3_plain_rewritable_disk:/$ ls
> s3_plain_rewritable_disk:/$ ls /fo
> s3_plain_rewritable_disk:/$ ls /foo
> s3_plain_rewritable_disk:/$ mkdir foo
> s3_plain_rewritable_disk:/$ ls
> foo
> s3_plain_rewritable_disk:/$ touch foo/bar
> s3_plain_rewritable_disk:/$ ls foo
> bar
> s3_plain_rewritable_disk:/$ touch foo/bar/bam/baz
> s3_plain_rewritable_disk:/$ ls foo
> bar
> s3_plain_rewritable_disk:/$ ls --recursive foo
> foo:
> bar
> ```

@azat To me, this is an issue with touch function for PlainRewritableObjectStorage.

Call stack (abridged):
1. touch
2. DiskObjectStorageTransaction::commit()
    - file is created to disk
    - metadata is updated in try and catch
4. MetadataStorageFromPlainObjectStorageWriteFileOperation::execute()
    - if addFile returns true then written = true else log event
6. InMemoryDirectoryPathMap::addFile()
    - if directory exists, add file to metadata and return true else false

In your example above, "touch /foo/bar/bam/baz" /foo/bar/bam directory do not exists and addFile() returns false.
 
Since execute() does not throw an exception when addFile returns false (directory doesn't exist), there is no error and the file does not exist in metadata. Note when this happens, the touch file does exist in the disk because commit() creates the file then updates metadata. There was no exception and therefore no rollback.

I think this could be solved by throwing an exception in execute() if addFile() returns false, but I have yet to test it out and I don't feel it should be done in this PR.

I've tested this in my PR and the master with the same results.

azat commented on src/Disks/ObjectStorages/InMemoryDirectoryPathMap.h:
```diff
@@ -210,22 +229,20 @@ class InMemoryDirectoryPathMap
     {
+        std::unordered_set<std::string> subdirectories;
         std::lock_guard lock(mutex);
         for (auto it = map.lower_bound(path); it != map.end(); ++it)
```
I've made the following changes in the test and found it

<details>

```diff
diff --git a/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.reference b/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.reference
index 8223b88edac..860427c7e89 100644
--- a/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.reference
+++ b/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.reference
@@ -21,3 +21,6 @@ Test ls with explicit path

 03600_disk_app_plain_rewritable_disks_list_cmd_default_hello/03600_disk_app_plain_rewritable_disks_list_cmd_default_world/03600_disk_app_plain_rewritable_disks_list_cmd_default_Clickhouse:

+Test no duplicated entries
+1
+2
diff --git a/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.sh b/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.sh
index e2e2bb05aa9..4dfe20596c5 100755
--- a/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.sh
+++ b/tests/queries/0_stateless/03600_disk_app_plain_rewritable_disks_list_cmd.sh
@@ -8,22 +8,39 @@ CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 config="$CUR_DIR/03600_disk_app_plain_rewritable_disks_list_cmd.xml"
 prefix="$CLICKHOUSE_TEST_UNIQUE_NAME"

-function test_copy()
+function test_ls()
 {
     disk_name="$1"

     clickhouse-disks -C "$config" --disk "$disk_name" --query "mkdir ${prefix}_hello; cd ${prefix}_hello; mkdir ${prefix}_world; cd ${prefix}_world; mkdir ${prefix}_Clickhouse;"
+
     printf "Test file can't be navigated" | clickhouse-disks -C "$config" --disk "$disk_name" --query "cd ${prefix}_hello; write ${prefix}_test_file.txt;"
     clickhouse-disks -C "$config" --disk "$disk_name" --query "cd ${prefix}_hello; read ${prefix}_test_file.txt;"
     clickhouse-disks -C "$config" --disk "$disk_name" --query "cd ${prefix}_hello; cd ${prefix}_test_file.txt;"  2>&1 | grep -o "BAD_ARGUMENTS"
+
     echo "Test can't navigate partial path"
     clickhouse-disks -C "$config" --disk "$disk_name" --query "cd ${prefix};"  2>&1 | grep -o "BAD_ARGUMENTS"
+
     echo "Test ls with implicit path"
     clickhouse-disks -C "$config" --disk "$disk_name" --query "cd ${prefix}_hello; ls; ls --recursive;"
+
     echo "Test ls with explicit path"
     clickhouse-disks -C "$config" --disk "$disk_name" --query "ls ${prefix}_hello/${prefix}_world; ls --recursive ${prefix}_hello/${prefix}_world"
+
+    echo "Test no duplicated entries"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "mkdir ${prefix}_dups"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "mkdir ${prefix}_dups/1/100"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "mkdir ${prefix}_dups/2/200"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "touch ${prefix}_dups/1/100/index"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "touch ${prefix}_dups/2/200/index"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "touch ${prefix}_dups/2/200/index.html"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "ls ${prefix}_dups"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "ls --recursive ${prefix}_dups"
+
+    # Cleanup
     clickhouse-disks -C "$config" --disk "$disk_name" --query "rm -r /${prefix}_hello/"
+    clickhouse-disks -C "$config" --disk "$disk_name" --query "rm -r /${prefix}_dups/"
 }

-test_copy "test_plain_rewritable_metadata_object_storage"
+test_ls "test_plain_rewritable_metadata_object_storage"

```

</details>

azat committed:
Merge remote-tracking branch 'u/master' into HEAD

All bugs has been fixed in https://github.com/ClickHouse/ClickHouse/pull/89125
